src/main/java/example/JaccardSimilarityMeasureExample.java

```diff
@@ -0,0 +1,173 @@
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.graph.Edge;
+import org.apache.flink.graph.Graph;
+import org.apache.flink.graph.Vertex;
+import org.apache.flink.types.NullValue;
+import org.apache.flink.util.Collector;
+
+import java.util.HashSet;
+import java.util.Iterator;
+
+public class JaccardSimilarityMeasureExample implements ProgramDescription {
+
+	public static void main(String [] args) throws Exception {
+
+		if (!parseParameters(args)) {
+			return;
+		}
+
+		ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
+
+		DataSet<Edge<Long, Double>> edges = getEdgesDataSet(env);
+
+		Graph<Long, NullValue, Double> graph = Graph.fromDataSet(edges, env);
+		// undirect the graph
+		Graph<Long, NullValue, Double> undirectedGraph = graph.getUndirected();
+
+		DataSet<Vertex<Long, HashSet<Long>>> verticesWithNeighbors = assignNeighborsAsValues(undirectedGraph.getEdges());
+
+		DataSet<Edge<Long, Double>> result = computeJaccardSimilarity(verticesWithNeighbors);
+
+		// emit result
+		if (fileOutput) {
+			result.writeAsCsv(outputPath, "\n", ",");
+		} else {
+			result.print();
+		}
+
+		env.execute("Executing Jaccard Similarity Measure");
+	}
+
+	@Override
+	public String getDescription() {
+		return "Vertex Jaccard Similarity Measure";
+	}
+
+	/**
+	 * Each vertex will have a HashSet containing its neighbors as value.
+	 *
+	 * @param edges
+	 * @return
+	 */
+	private static DataSet<Vertex<Long, HashSet<Long>>> assignNeighborsAsValues(DataSet<Edge<Long, Double>> edges) {
+
+		return edges.groupBy(0).reduceGroup(new GroupReduceFunction<Edge<Long, Double>, Vertex<Long, HashSet<Long>>>() {
+
+			@Override
+			public void reduce(Iterable<Edge<Long, Double>> iterableEdge,
+							   Collector<Vertex<Long, HashSet<Long>>> collector) throws Exception {
+
+				Iterator<Edge<Long, Double>> iteratorEdge = iterableEdge.iterator();
+				HashSet<Long> neighbors = new HashSet<Long>();
+				Long vertexSrcId = 0L;
+
+				while(iteratorEdge.hasNext()) {
+					Edge<Long, Double> iteratorEdgeNext = iteratorEdge.next();
+					vertexSrcId = iteratorEdgeNext.getSource();
+					neighbors.add(iteratorEdgeNext.getTarget());
+				}
+
+				collector.collect(new Vertex<Long, HashSet<Long>>(vertexSrcId, neighbors));
+			}
+		});
+	}
+
+	/**
+	 * Consider the edge x-y
+	 * We denote by sizeX and sizeY, the neighbors hash set size of x and y respectively.
+	 * sizeX+sizeY = union + intersection of neighborhoods
+	 * size(hashSetX.addAll(hashSetY)).distinct = union of neighborhoods
+	 * The intersection can then be deduced.
+	 *
+	 * Jaccard Similarity is then, the intersection/union.
+	 *
+	 * @param verticesWithNeighbors
+	 * @return
+	 */
+	private static DataSet<Edge<Long, Double>> computeJaccardSimilarity(
+			DataSet<Vertex<Long, HashSet<Long>>> verticesWithNeighbors) {
+
+		return verticesWithNeighbors.cross(verticesWithNeighbors).with(new CrossFunction<Vertex<Long, HashSet<Long>>,
```

Comment:
why would you do a cross here? this is way too expensive.. you can get each node's neighbors from the graph.. that's the whole point of using a gelly :)